### PR TITLE
HTTP broadcast adapter

### PR DESCRIPTION
**Pull request by palkan**
### What is the purpose of this pull request?

Add support for HTTP broadcasting (see https://github.com/anycable/anycable/pull/100).

### What changes did you make? (overview)

- [x] **Added HTTP pub/sub adapter**
- [x] Refactored the way we attach new handlers to HTTP server or create new servers. The `server.ForPort(port)` function is introduced which creates a new server for the specified port or returns the existing one (Host and SSL options are shared among all the servers created this way).
- [x] Added some missing HTTP handlers tests.
- [x] Refactored `node` mocks to make it easier to use them in other packages tests.

### Is there anything you'd like reviewers to focus on?

A couple of things I'd be glad to get some feedback on:
- `server.allServers` trick to re-use HTTP servers for different components
- `node.AppNode` interface and its usages; it's a very basic implementation useful just for pub/sub components; I gave up on adding it everywhere—too much work; maybe, later.

### Checklist

- [x] I've added tests for this change
- [x] I've added a Changelog entry
- [x] I've updated documentation https://github.com/anycable/docs.anycable.io/pull/26

### CODE REVIEW

File: pubsub/http.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+// Start creates an HTTP server or attaches a handler to the existing one
+func (s *HTTPSubscriber) Start() error {
+	server, err := server.ForPort(strconv.Itoa(s.port))
+
+	if err != nil {
+		return err
+	}
+
+	s.server = server
+	s.server.Title = "Pub/Sub HTTP server"
```
Comment (sponomarev) [inline pubsub/http.go:65]: Would it be a race condition to mutate title for an existent server? It should handle gracefully a newly created one but what if we attach to the existent?

File: cmd/anycable-go/main.go [hunk 2]
```diff
@@ -93,17 +99,37 @@ func main() {
 		}
 	}()
 
-	wsServer := buildServer(config.Host, strconv.Itoa(config.Port), &config.SSL)
-	wsServer.Mux.Handle(config.Path, server.WebsocketHandler(appNode, config.Headers, &config.WS))
+	server.SSL = &config.SSL
+	server.Host = config.Host
+
+	wsServer, err := server.ForPort(strconv.Itoa(config.Port))
```
Comment (sponomarev) [inline cmd/anycable-go/main.go:105]: Should we set server title here as well?

File: server/server.go [hunk 3]
```diff
@@ -14,24 +14,48 @@ import (
...
+	allServers map[string]*HTTPServer = make(map[string]*HTTPServer)
+	// Host is a default bind address for HTTP servers
+	Host string = "localhost"
+	// SSL is a default configuration for HTTP servers
+	SSL *SSLConfig
+)
+
+// ForPort creates new or returns the existing server for the specified port
+func ForPort(port string) (*HTTPServer, error) {
+	if _, ok := allServers[port]; !ok {
```
Comment (sponomarev) [inline server/server.go:40]: The idea of a reusable server pool is great! 👍

File: pubsub/http.go (hunk 1 above)
Comment (palkan) [reply pubsub/http.go:65]: Good point. Probably, we should remove this hack in favor of in-place logging for every server (the only thing we use to for now).

File: node/ws_handler.go [hunk 4]
```diff
@@ -41,6 +42,11 @@ func WebsocketHandler(app *Node, fetchHeaders []string, config *WSConfig) http.H
 		}
 
 		url := r.URL.String()
+
+		if !r.URL.IsAbs() {
```
Comment (sponomarev) [inline node/ws_handler.go:46]: 👍

File: pubsub/http.go [hunk 5]
```diff
@@ -62,14 +62,13 @@ func (s *HTTPSubscriber) Start() error {
 	}
 
 	s.server = server
-	s.server.Title = "Pub/Sub HTTP server"
 	s.server.Mux.Handle(s.path, http.HandlerFunc(s.Handler))
 
 	s.log.Infof("Accept broadcast requests at %s%s", s.server.Address(), s.path)
 
-	if err := s.server.Start(); err != nil {
+	if err := s.server.StartAndAnnounce("Pub/sub server"); err != nil {
```
Comment (sponomarev) [inline pubsub/http.go:69]:
```suggestion
	if err := s.server.StartAndAnnounce("Pub/Sub HTTP server"); err != nil {
```

File: server/server.go [hunk 6]
```diff
@@ -81,21 +78,37 @@ func NewServer(host string, port string, ssl *SSLConfig) (*HTTPServer, error) {
...
 	if s.secured {
-		s.log.Infof("Starting %s with SSL enabled at %v", s.Title, s.addr)
 		return s.server.ListenAndServeTLS("", "")
 	}
 
-	s.log.Infof("Starting %s at %v", s.Title, s.addr)
 	return s.server.ListenAndServe()
 }
 
+// StartAndAnnounce prints server info and start server
```
Comment (sponomarev) [inline server/server.go:97]:
```suggestion
// StartAndAnnounce prints server info and starts server
```

**Pull request merged**